### PR TITLE
Revise layout of group convolution

### PR DESCRIPTION
**Pull request by rocking5566**
Follow the definition of group convolution
User might have NHWGC/GKYXC/NHWGK (x, weight, y) in memory space
Revise the layout of standalond conv2D and conv2D quantization

### CODE REVIEW

File: client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp [hunk 1]
```diff
@@ -52,50 +52,24 @@ struct SimpleDeviceMem
...
-    std::rotate(
-        rbegin(wei_strides), std::next(rbegin(wei_strides)), std::next(rbegin(wei_strides), 3));
-    std::rotate(
-        rbegin(out_lengths), std::next(rbegin(out_lengths)), std::next(rbegin(out_lengths), 3));
-    std::rotate(
-        rbegin(out_strides), std::next(rbegin(out_strides)), std::next(rbegin(out_strides), 3));
+    // We have NHWGC/GKYXC/NHWGK (x, weight, y) in memory space
+    // However, CK's API only accept length and stride with order of GNCHW/GKCYX/GNCHW
+    // Hence, we need to adjust the order of stride
+    std::array<ck::index_t, 5> in_lengths{G, N, C, Hi, Wi};
```
Comment (iq136boy) [inline client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp:58]: Should it be in_lengths {G, N, C1, Hi, Wi} where C1=C/G?

File: client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp [hunk 2]
```diff
@@ -52,50 +52,24 @@ struct SimpleDeviceMem
...
+    std::array<ck::index_t, 5> in_strides{C, Hi * Wi * G * C, 1, Wi * G * C, G * C};
+    std::array<ck::index_t, 5> wei_lengths{G, K, C, Y, X};
```
Comment (iq136boy) [inline client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp:60]: Same here C should be replaced by C1=  C/G, right?

File: client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp [hunk 3]
```diff
@@ -52,50 +52,24 @@ struct SimpleDeviceMem
...
+    std::array<ck::index_t, 5> out_lengths{G, N, K, Ho, Wo};
+    std::array<ck::index_t, 5> out_strides{C, Ho * Wo * G * C, 1, Wo * G * C, G * C};
 
     std::array<ck::index_t, NumDimSpatial> filter_strides{1, 1};
     std::array<ck::index_t, NumDimSpatial> filter_dilations{1, 1};
     std::array<ck::index_t, NumDimSpatial> input_left_pads{1, 1};
     std::array<ck::index_t, NumDimSpatial> input_right_pads{1, 1};
 
-    SimpleDeviceMem in(sizeof(InDataType) * G * N * Hi * Wi * C);
+    SimpleDeviceMem in(sizeof(InDataType) * N * Hi * Wi * G * C);
```
Comment (iq136boy) [inline client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp:70]: Also here C should be C/G.

**Review by iq136boy — CHANGES_REQUESTED** [review]

File: client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp (hunk 1 above)
Comment (rocking5566) [reply client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp:58]: I add the comment to describe the meaning of each parameter in 6aeeb31af581fb183088902298670db10e3143c1
C means input channel per group now

File: client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp (hunk 2 above)
Comment (rocking5566) [reply client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp:60]: I add the comment to describe the meaning of each parameter in 6aeeb31af581fb183088902298670db10e3143c1
C means input channel per group now

File: client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp (hunk 3 above)
Comment (rocking5566) [reply client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp:70]: Same as before

**Comment by zjing14** [comment]
@iq136boy Please review the PR again and approve it if the PR meets your requirement.

File: client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp (hunk 2 above)
Comment (iq136boy) [inline client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp:60]: What about K? Also means K per group?

**Review by iq136boy — CHANGES_REQUESTED** [review]

File: client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp (hunk 2 above)
Comment (rocking5566) [reply client_example/07_grouped_convnd_fwd/grouped_conv2d_fwd.cpp:60]: Revise to output channel per group now  e711cfdbb72df2e1b26e67e7dabc921ae32546fa

**Review by iq136boy — APPROVED** [review]

**Review by iq136boy — CHANGES_REQUESTED** [review]
Please ignore this comment.

**Review by iq136boy — APPROVED** [review]

**Pull request merged**